### DISCUSSION TITLE
Scheduled Export to Cloud Resource

### DISCUSSION THREAD

**Discussion by jringbox** [tmo1/sms-ie#215] [Q&A] [answered]
With sms-ie, I have setup my "Export to directory" to "My Documents" folder.  This works as expected for all things that are exported by when providing the time each day for this to occur.

My issue is how to get this to sync with a cloud resource, specifically Google Drive.

From what I can tell, I would have to use a third-party tool like FolderSync to sync files from my android, esp in the case of Google Drive.  At one time Google Drive could do specific folders for syncing but that feature was taken away.  tmo1 believe there should be a way to do this without FolderSync since sms-ie does all input and output via the Android Storage Access Framework (SAF).  I guess my expectation with this in mind in that when I click on "Export to directory", I would see other resources I can export to, like "Google Drive".  But all I see are internal storage locations.

Has anyone found a way to do this?  Has anyone synced to a cloud resource?  How did you setup your device so the cloud resource appears?

**Comment by tmo1** [comment]
>  tmo1 believe there should be a way to do this without FolderSync since sms-ie does all input and output via the Android Storage Access Framework (SAF). I guess my expectation with this in mind in that when I click on "Export to directory", I would see other resources I can export to, like "Google Drive". But all I see are internal storage locations.

Accessing Google Drive via the SAF apparently requires the Google Drive app - see [here](https://news.ycombinator.com/item?id=19509837). Do you have it installed?

**Reply by jringbox** [reply]
I know.  I expected the "Export to directory" to show other sources like "Google Drive" but there was nothing.  I went through my internal folder and looked for options to make a folder a cloud resource to Google Drive and I couldn't.

I do have Google Drive installed.

When I look online I see many people frustrated that Google Drive won't share from their device and they have to install FolderSync.

Also, when I use the desktop version of Google Drive for my account, I click on the hamburger for a folder and go to "Open with".  Under this option is "Connect more apps".  There is a Google Workspace Marketplace with apps that tie into Google Drive for accessibility apparently.  So maybe cloud sharing using SAF is not that simple with Google Drive?

**Comment by tmo1** [comment]
I don't know what to tell you. You can try other apps that use the SAF interface and see if they can see your Google Drive account. If they can, then perhaps I can troubleshoot further why SMS I/E can't, but if they can't, then the problem is with Google Drive or your account / configuration, and I can't troubleshoot that for you.

**Comment by jringbox** [comment]
I searched online and struggled trying to find a good answer.  Then, I tried a different approach and searched on Google with "how to get android app to access google drive" and got this result in the first paragraph from the Google AI response.

"To allow an Android app to access your Google Drive, you need to grant the app permission to access your Google account within the app settings; this usually involves a prompt within the app asking you to sign in with your Google account, which will then give the app the necessary permissions to read and write files on your Google Drive."

On one of the links to the right associated to this AI result, I went to "Allow third-party apps for Drive files".

https://support.google.com/a/answer/6105699?hl=en

When I read this there was a blurb about "Allow users to access Google Drive with the Drive SDK API."  I went ahead and looked up "Drive SDK API" at this link.

https://developers.google.com/drive/api/guides/about-sdk

The diagram on this page shows that through oAuth one can access Google Drive folders such as "My Drive" and other "Shared drives".  Apparently, Android apps need to login to the cloud storage, and authorization needs to be provided for the app by the user to be able to access Google Drive.  This is the best answer I found to why the app is not able to access Google Drive.

This probably is out of the scope of your coding priorities and all of this doesn't mention SAF to any extent.  So I think this is non-SAF approach which goes against your coding approach.

I tried looking for a SAF app but I could not figure out what app was "exclusive" to SAF.  Amaze app uses SAF but I don't know to what extent.  I was able to get to Google Drive using the app within a certain context.  There were Google Drive apps but that doesn't mean they use SAF exclusively.

Anyway, when time allows, I'm going to keep searching and see if I can find another solution or an exclusive SAF app.

**Comment by tmo1** [comment]
>https://support.google.com/a/answer/6105699?hl=en

> When I read this there was a blurb about "Allow users to access Google Drive with the Drive SDK API." I went ahead and looked up "Drive SDK API" at this link.

> https://developers.google.com/drive/api/guides/about-sdk

> The diagram on this page shows that through oAuth one can access Google Drive folders such as "My Drive" and other "Shared drives". Apparently, Android apps need to login to the cloud storage, and authorization needs to be provided for the app by the user to be able to access Google Drive. This is the best answer I found to why the app is not able to access Google Drive.

Google apparently makes Drive accessible via a special API, which apps can use to access Drive directly, without going through the generic SAF framework.

> The diagram on this page shows that through oAuth one can access Google Drive folders such as "My Drive" and other "Shared drives". Apparently, Android apps need to login to the cloud storage, and authorization needs to be provided for the app by the user to be able to access Google Drive. This is the best answer I found to why the app is not able to access Google Drive.

My understanding is that in addition to its proprietary API, Drive should also be available via the generic SAF. I suggest you ask Google itself, or ask on a third-party Google forum, for help regarding this.

> This probably is out of the scope of your coding priorities and all of this doesn't mention SAF to any extent. So I think this is non-SAF approach which goes against your coding approach.

Correct. I'm sorry, but I'm not going to devote considerable time and effort to coding for Google's proprietary API where the SAF approach generally works fine.

> Anyway, when time allows, I'm going to keep searching and see if I can find another solution or an exclusive SAF app.

You can try [Material Files](https://github.com/zhanghai/MaterialFiles) - I'm not sure, but I don't see any mention of Google Drive integration.

**Comment by jringbox** [comment]
Here's a question.  In sms-ie, if I go to settings and click on "Export to directory:", I see this at the top.

![SMS-IE-clip](https://github.com/user-attachments/assets/b524674d-3deb-4582-b4a4-3bbe3acf3543)

Material Files is completely different application, but it has a hamburger menu at the top left as shown below.

![Material-clip](https://github.com/user-attachments/assets/268cc65c-7d2c-49b9-b7c7-a9b7229e4910)

If I click on the hamburger menu, I see this.

![Material-Menu-clip](https://github.com/user-attachments/assets/e1e23bd3-7a13-4f48-ad79-a2206713b719)

From this top-level menu, I see "Drive" listed and I can go to it.  Somehow this app is able to access it and sms-ie is not.  I'm not sure if they are using SAF or not.  I guess that is the magic question I need to ask them?  I guess I'll ask them.

**Comment by tmo1** [comment]
> From this top-level menu, I see "Drive" listed and I can go to it. Somehow this app is able to access it and sms-ie is not. I'm not sure if they are using SAF or not. I guess that is the magic question I need to ask them? I guess I'll ask them.

Hm. There have been numerous requests for Material Files Google Drive integration: [#95](https://github.com/zhanghai/MaterialFiles/issues/95), [#541](https://github.com/zhanghai/MaterialFiles/issues/541), [#619](https://github.com/zhanghai/MaterialFiles/issues/619), [#705](https://github.com/zhanghai/MaterialFiles/issues/705), [#715](https://github.com/zhanghai/MaterialFiles/issues/715), [#763](https://github.com/zhanghai/MaterialFiles/issues/763), [#925](https://github.com/zhanghai/MaterialFiles/issues/925), [#1070](https://github.com/zhanghai/MaterialFiles/issues/1070), [#1304](https://github.com/zhanghai/MaterialFiles/issues/1304). I can't find where such support was added, but the comments in 95 as well as 619 suggest that @zhanghai was looking at implementing the Drive API.

**Comment by jringbox** [comment]
This is interesting.  I didn't post everything I thought I did when the Google Drive appeared.  What I did was go to "Internal Storage" and choose "Android/obb".

![Material-Android-obb](https://github.com/user-attachments/assets/8b7e0606-ce56-46a1-921f-451b5ccd9c2c)

Then I got this on my menu.

![Material-Android-obb-menu](https://github.com/user-attachments/assets/abc9ebd1-3488-4f2f-8f4d-3af9404c2c00)

From here, I clicked on "Android/obb" and the new navigation menu appears with Google Drive.  Not sure what that is doing.  No idea why obb has a connection to Drive?!?  However, that is probably all Drive API logic as you mentioned and out of scope.  It's very interesting how it can flip-flop from what appears to be a SAF navigation? to navigation menu with Drive API menu navigation.  I'm not sure what the developer is doing.  I did sent him an e-mail though.  The interesting thing is that his webpage says he is a Staff Software Engineer at Google.

Probably wasting your time with all this info but I do find it interesting.

In the e-mail, I did point him to this post.  He might surprise us with some insight.

**Comment by jringbox** [comment]
Here's an article about basic SAF implementation.

https://medium.com/swlh/sample-for-android-storage-access-framework-aka-scoped-storage-for-basic-use-cases-3ee4fee404fc

I looked at the code in AndroidManifest.xml and it doesn't have `android.permission.WRITE_EXTERNAL_STORAGE` which is good, not sure about all the other permissions though.  But I see in build.gradle that this is missing.

`androidx.documentfile:documentfile:1.0.1`

I think `MainActivity.kt` is missing `resultCode == RESULT_OK && requestCode == REQUEST_CODE` check based on the article.

And the rest, I'm not sure where and how they should be implemented.  I'm getting the impression SAF isn't fully implemented based on this article.  But I'm not an Android developer so I'm not sure.  I couldn't find references to some of his implementation in the code.  However, maybe I didn't search right or maybe it's implemented differently than this article.  Maybe someone can point me to where the SAF logic is implemented?

**Comment by tmo1** [comment]
>  But I see in build.gradle that this is missing.

> `androidx.documentfile:documentfile:1.0.1`

It's unnecessary, since it's a transitive dependency of `com.google.android.material:material` - see [here](https://stackoverflow.com/questions/59861038/why-do-i-need-to-add-material-theme-dependency-in-order-to-resolve-documentfile).

> I think MainActivity.kt is missing resultCode == RESULT_OK && requestCode == REQUEST_CODE check based on the article.

[It's there](https://github.com/tmo1/sms-ie/blob/e23f01540920bbad9b41c4187f8be40410946301/app/src/main/java/com/github/tmo1/sms_ie/MainActivity.kt#L342).

>  I'm getting the impression SAF isn't fully implemented based on this article.  But I'm not an Android developer so I'm not sure.

As far as I know, SAF support is properly implemented.

**Reply by jringbox** [reply]
> > I think MainActivity.kt is missing resultCode == RESULT_OK && requestCode == REQUEST_CODE check based on the article.
> 
> [It's there](https://github.com/tmo1/sms-ie/blob/e23f01540920bbad9b41c4187f8be40410946301/app/src/main/java/com/github/tmo1/sms_ie/MainActivity.kt#L342).
> 
I don't know Kotlin so this is all out of my programming experience and time though I do know Java.  I saw all those checks for perms in `MainActivity.kt`.  What I didn't know is if literally `requestCode == REQUEST_CODE` check was needed.  When I looked at his code, I saw him handling the requests very differently than sms-ie did.

https://github.com/PavelBorze/SAF_sample/blob/main/app/src/main/java/com/axanor/saf_sample/MainActivity.kt

His class implementation is different but he's not dealing with any specific business requirements for his demo app.  sms-ie has message, contacts, etc stuff going on.  However, his `onCreate` calls an `openDocumentTree` method he has implemented and then he calls `makeDoc` when needed and goes back and forth.  This appears to be his directory navigation workflow.  I think what his `MainActivity.kt` does is what sms-ie `SettingsActivity.kt` is accomplishing for finding export directory location.

When I have time (whenever that is), I guess I'll have to check his project out and see if I have an IDE that can compile it and generate an apk, most likely Android Studio will build it for me.  I just want to see if his app shows my cloud resource or not on my phone.  Also, I'll have to read and hunt around some more on my SAF issue when I have time.  If I get this far, I might be able to checkout sms-ie code and play around with it but I I'll have to see.

**Reply by jringbox** [reply]
As fyi, I did get that sample SAF project built from the medium.com article and it works no different than sms-ie.  I guess I must be doing something wrong on how my phone is setup or something.  In fact, both android phones I have here don't access the resources I want.  I'll have to research some more to figure out where things aren't working for me.

**Reply by jringbox** [reply]
I got a little irritated that Google Drive would not show up.  So I decided to click on something else.  I went back to the main app page and clicked on Export Messages.  Instead of navigating the folders from where I was, I clicked on the hamburger and now I see this.

![image](https://github.com/user-attachments/assets/d731d95e-fcd3-40b4-8d56-caafbe915a0e)

What the heck?!?  How did I miss this?!?  Then I safely pursued to Import Contacts and I see this.

![image](https://github.com/user-attachments/assets/a4ef4d34-6a3e-4b6f-ae3e-64d591ff8dd6)

Okay.  So the cloud resource is showing up for buttons on the main page.  But why aren't the cloud resources showing up under Setting->Export Directory?!?  Clearly sms-ie is capable and just isn't doing it under this context.

**Comment by tmo1** [comment]
> Okay. So the cloud resource is showing up for buttons on the main page. But why aren't the cloud resources showing up under Setting->Export Directory?!? Clearly sms-ie is capable and just isn't doing it under this context.

I tested using a Nextcloud account on my phone, and the Nextcloud account correctly shows up in the contexts of both manual operations ("buttons on the main page") as well as scheduled export (when selecting "Export to directory"). I don't know why Google Drive isn't showing up for you in the latter context.

**Comment by srhaider000** [comment]
Apologies in advance for being unable to offer any insight into this, but just to add one more datapoint I'm experiencing the exact same issue attempting to export to proton drive. Doing a manual, one time export yields the file explorer with a hamburger menu that allows me to select Proton Drive, but trying to setup periodic backups only lets me select from my phone's internal storage. 

In case it's useful, I'm on a Pixel 6a running GrapheneOS.

**Comment by tmo1** [comment, marked as answer]
> Apologies in advance for being unable to offer any insight into this, but just to add one more datapoint I'm experiencing the exact same issue attempting to export to proton drive.

Your data point was the impetus for me to finally figure out what's going on here. When SMS I/E performs manual exports, it uses single-file API calls that are widely supported by cloud SAF providers, but when scheduling exports, the app needs a folder / directory into which it can repeatedly write export files, so it uses [`ACTION_OPEN_DOCUMENT_TREE`](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT_TREE), which [many](https://stackoverflow.com/questions/56539920/storage-access-framework-doesnt-show-cloud-roots-when-using-action-open-documen) [cloud](https://stackoverflow.com/questions/56664638/storage-access-framework-does-the-file-picker-app-not-implement-action-open-do) [SAF](https://medium.com/steadfast-innovation/open-letter-to-cloud-storage-apps-on-android-a2d0eb919fc4) providers apparently do **not** support - including (rather surprisingly) Google Drive (see [here](https://issuetracker.google.com/issues/135636079), [here](https://issuetracker.google.com/issues/139965899), and [here](https://issuetracker.google.com/issues/65673235)) and Proton Drive. Other Android apps have encountered this same problem: [Aegis Authenticator](https://github.com/beemdevelopment/Aegis/issues/507); [SeedVault](https://github.com/seedvault-app/seedvault/issues/432) (and see [here](https://github.com/seedvault-app/seedvault/issues/167)).

An effective if somewhat cumbersome solution is the one eventually adopted by SeedVault: use an Android rclone wrapper that makes cloud storage facilities available through the Android SAF (with [`ACTION_OPEN_DOCUMENT_TREE`] support; [SeedVault went with RoundSync](https://github.com/seedvault-app/seedvault/pull/609), which I know nothing about, but I can strongly recommend [RSAF](https://github.com/chenxiaolong/RSAF), whose author is a major contributor to SMS I/E.

**Reply by chenxiaolong** [reply]
This was one of the main reasons I made RSAF. At the time, it seemed like [cifs-documents-provider](https://github.com/wa2c/cifs-documents-provider), NextCloud, and a couple other open source SAF providers were the only ones to support `ACTION_OPEN_DOCUMENT_TREE`.

Sad to see the situation hasn't really changed in the past couple years with the major cloud providers.

**Reply by jringbox** [reply]
So where does that leave my situation with Google Drive?  I looked at RSAF but I'm not sure on it's usage.  Are we saying I can install that, configure it and solve my issue with sms-ie?

**Reply by srhaider000** [reply]
> Are we saying I can install that, configure it and solve my issue with sms-ie?

Yes, install the app and add a new remote and then follow the directions [here](https://rclone.org/drive/). Once you have set that up you should see the slide out menu in the storage menu with Google drive as an option.
